packages/kbn-openapi-bundler/src/bundler/document_processors/reduce_all_of_items/flatten_folded_all_of_items.ts

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { DocumentNodeProcessor } from '../../types';
+
+/**
+ * Creates a node processor to flatten folded allOf items.
+ *
+ * Folded `allOf` schemas is usually a result of inlining references.
+ *
+ * Example:
+ *
+ * The following folded `allOf`s
+ *
+ * ```yaml
+ * allOf:
+ *   - allOf:
+ *     - type: object
+ *       properties:
+ *         fieldA:
+ *           $ref: '#/components/schemas/FieldA'
+ *     - type: object
+ *       properties:
+ *         fieldB:
+ *           type: string
+ * ```
+ *
+ * will be transformed to
+ *
+ * ```yaml
+ * allOf:
+ *   - type: object
+ *     properties:
+ *       fieldA:
+ *         $ref: '#/components/schemas/FieldA'
+ *   - type: object
+ *     properties:
+ *       fieldB:
+ *         type: string
+ * ```
+ *
+ */
+export function createFlattenFoldedAllOfItemsProcessor(): DocumentNodeProcessor {
+  return {
+    leave(node) {
```

Comment:
There are two phased at which OpenAPI document is traversed.

- At the first phase diving from root to leaves happens (like event's capture phase in DOM). It means we `enter` a subtree. Processing isn't happening here so it's convenient to analyze not yet processed nodes. For example understand which nodes should be inlined.
- At the second phase bubbling from leaves to root happens (like event's bubble phase in DOM). It means we `leave` s subtree. It's possible to modify a node during this phase while it's not necessary.

This naming highlights the flexibility. `enter` and `leave` methods could be used to collect metrics or analyze the document. Current implementation doesn't have such functionality but it could be added without any effort.

I extended a comment for `DocumentNodeProcessor` in `packages/kbn-openapi-bundler/src/bundler/types.ts`. While we don't expose DocumentNodeProcessor API I think package README shouldn't be updated.